src/lib/core/error_payload.c

```diff
@@ -38,27 +37,30 @@ error_payload_prepare(struct error_payload *e, const char *name,
 		      uint32_t value_size, uint32_t extra)
 {
 	struct error_field *f;
-	uint32_t name_size = strlen(name) + 1;
-	uint32_t data_offset = offsetof(struct error_field, name[name_size]);
-	uint32_t total = data_offset + value_size + extra;
-
 	struct error_field **fields = e->fields;
+	struct grp_alloc all = grp_alloc_initializer();
 	int count = e->count;
 	int i;
 	for (i = 0; i < count; ++i) {
 		f = fields[i];
 		if (strcmp(name, f->name) != 0)
 			continue;
-		f = xrealloc(f, total);
-		goto set;
+		TRASH(f);
+		free(f);
+		goto alloc_field;
```

Comment:
Here we can avoid reallocation if `value_size` and `extra` are unchanged, but it will require an additional `uint32_t extra` field in `struct error_field`.
The errors are supposed to be rare, so this optimization isn't worth it.